neutron-ha-tool: Add DHCP agent evacuation (SOC-11046)

skazi0 opened the pull request:
DHCP agent handles networks and ports in a way similar to L3 agent's
routers. The assignment of these to the network nodes need to be
properly orchestrated to keep things under control e.g. in upgrade
scenario where we need to move network agents out from node before
upgrading to make the process non-disruptive.

skazi0 committed:
neutron-ha-tool: Add DHCP agent evacuation (SOC-11046)

DHCP agent handles networks and ports in a way similar to L3 agent's
routers. The assignment of these to the network nodes need to be
properly orchestrated to keep things under control e.g. in upgrade
scenario where we need to move network agents out from node before
upgrading to make the process non-disruptive.

jsuchome commented on files/default/neutron-ha-tool.py:
```diff
@@ -576,6 +600,195 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+
+    :param qclient: A neutronclient
+    :param noop: Optional noop flag
+    :param agent_host: the hostname of the DHCP agent to migrate networks from
```
Could you add comments for all parameters?

skazi0 replied:
done

jsuchome commented on files/default/neutron-ha-tool.py:
```diff
@@ -576,6 +600,195 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+                                                exclude_agent=agent_to_evacuate)
+
+    if len(target_list) < 1:
+        LOG.error("There are no DHCP agents alive to migrate routers onto")
```
s/routers/networks

skazi0 replied:
done + fixed one log message in L3 migration code (separate commit)

jsuchome commented on files/default/neutron-ha-tool.py:
```diff
@@ -576,6 +600,195 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+def migrate_dhcp_network(qclient, network, agent, target,
+                         wait_for_network=True):
+    """
+    Returns nothing, and raises exceptions on errors.
```
That might be a good idea ... but for example `migrate_router_safely` is not using exceptions but return values.

It could be a good thing to have it consistent in the same tool...

skazi0 replied:
I'm not really sure what is the benefit of this additional "safely" wrapper function. The only additional functionality is skipping migration of live agents which is not included in DHCP evacuation. Besides that, I removed this additional nesting level and included the try-catch directly in `migrate_dhcp_networks_from_agent`.

jsuchome replied:
I'm just pointing out to the different way dhcp_ and router_ methods are reporting errors. It would be nicer if ti were consistent

rhafer replied:
> I'm not really sure what is the benefit of this additional "safely" wrapper function. 

Me neither. But I assume it was done for readablity reason. See https://github.com/SUSE-Cloud/cookbook-openstack-network/commit/41a048c5325c62da8207adcfd5bc531ee57b323a

Not sure if the dhcp functions reached that state for complexity yet.

jsuchome commented on files/default/neutron-ha-tool.py:
```diff
@@ -576,6 +600,195 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+        qclient.remove_network_from_dhcp_agent(agent['id'], network['id'])
+        LOG.debug("Removed network from agent=%s" % agent['id'])
+
+        # ensure it is removed
```
So `remove_network_from_dhcp_agent` does not return anything reasonable so that we have to check the result by another call?

skazi0 replied:
I followed the pattern from routers migration here. As far as I can see in `neutronclient` code, these functions are just thin wrappers around API calls and return values represent the HTTP return codes, so I guess it can only return something like "the request was accepted" but not "the network was actually removed from agent".

skazi0 committed:
neutron-ha-tool: Fix log message (trivial)

The log message about failed "add" operation listed source agent
instead of target one.

jsuchome commented on files/default/neutron-ha-tool.py:
```diff
@@ -576,6 +600,199 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+
+        # TODO: what to do if target already has the network?
+        # skip? pick another target? remote and re-add?
+        if not dhcp_network_is_on_agent(qclient, network['id'], target['id']):
```
Why the check again? In this part of code `dhcp_network_is_on_agent` should already return false.... or do you think something could have happened during `wait_dhcp_network_migrated`?

jsuchome commented on files/default/neutron-ha-tool.py:
```diff
@@ -576,6 +600,199 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+
+    LOG.info("Migrating network=%s from agent=%s to agent=%s",
+             network['id'], agent['id'], target['id'])
+
```
`network['id'], agent['id'], target['id']` are used several times in this function, maybe save them into variables?

skazi0 replied:
done.

jsuchome commented on files/default/neutron-ha-tool.py:
```diff
@@ -576,6 +600,194 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+    with term_disabled():
+
+        # network might have been removed in the mean time
+        if dhcp_network_is_on_agent(qclient, network['id'], target['id']):
```
I think here you need `agent['id']`. (also, see my older comment about using variables for those ids...)

skazi0 replied:
true

jsuchome commented on files/default/neutron-ha-tool.py:
```diff
@@ -576,6 +600,200 @@ def l3_agent_evacuate(qclient, agent_host, agent_picker, router_filter,
+            qclient.remove_network_from_dhcp_agent(agent_id, network_id)
+            LOG.debug("Removed network from agent=%s" % agent_id)
+        # network might have been removed in the mean time
+        except NotFound as e:
```
I believe it's actually `ObjectNotFound`, see https://pastebin.nue.suse.com/25407/src (neutron-server log)

skazi0 replied:
fixed but not tested

jsuchome replied:
Sorry this was wrong

jsuchome reviewed (commented):
This seems to work correctly now. Could you squash the commits?
@rhafer would you find time for review?

skazi0 commented:
rebased and squashed

rhafer reviewed (commented):
I think this is mostly ok. But have you ever tested it with dhcp_agents_per_network set to some thing larger than 1 ?

1. In that case it might be ok to just remove the network from the source agent and not add it anywhere else during the upgrade (provided that there is at least one dhcp port up for that network).

2. I guess it should at least be checked whether the target node already has a dhcp port up for the specific network.

Take all this with a grain of salt and if you feel the current code is "good enough (TM)" feel free to merge it. After all I don't have much visibility anymore into the project.

jgrassler reviewed (approved):
Looks good as far as I can tell from reading through it (haven't tested it in the half hour I spent on this, obviously). Between, that and you testing it to destruction the past couple of days, my verdict is a +1.

On an unrelated note: I know that code has been with us for a good, long while...has it been with us since the time when Neutron was still named Quantum? The use of the `qclient` variable name certainly would suggest so...
